keyshare.go

```diff
@@ -1,12 +1,208 @@
+		keyID := keyID
+		keyIDs[publicKeyIdentifier{issuer: key.Issuer, counter: key.Counter}] = &keyID
+	}
+
+	for _, builder := range builders {
+		c, err := builder.Commit(randomizers)
+		if err != nil {
+			return KeyshareCommitmentRequest{}, nil, err
+		}
+		var otherComms []*big.Int
+		if len(c) > 2 {
+			otherComms = c[2:]
+		}
+		pk := builder.PublicKey()
+		hashInput = append(hashInput, KeyshareUserChallengeInput[T]{
+			KeyID:            keyIDs[publicKeyIdentifier{issuer: pk.Issuer, counter: pk.Counter}],
+			Value:            new(big.Int).Set(c[0]),
+			Commitment:       new(big.Int).Set(c[1]),
+			OtherCommitments: otherComms,
+		})
+	}
+
+	bts, err := keyshareUserCommitmentsHash(hashInput)
+	if err != nil {
+		return KeyshareCommitmentRequest{}, nil, err
+	}
+	return KeyshareCommitmentRequest{HashedUserCommitments: bts}, hashInput, nil
+}
+
+// KeyshareUserResponseRequest computes the user's second message to the keyshare server in the
+// keyshare protocol, containing its response in the joint computation of the zero- knowledgeproof
+// of the secret key. Also returns the challenge to be used in constructing the proofs.
+func KeyshareUserResponseRequest[T comparable](
+	builders ProofBuilderList,
+	randomizers map[string]*big.Int,
+	hashInput []KeyshareUserChallengeInput[T],
+	context, nonce *big.Int,
+	signature bool,
+) (KeyshareResponseRequest[T], *big.Int, error) {
+	// Extract the user secret from the builders. Since this secret will be the same across all
+	// builders, we can just take it off the first one.
+	// (We extract it manually like this instead of adding a method to the ProofBuilder interface,
+	// because we don't want to expose a method to retrieve the secret in the gabi public API.)
+	var userSecret *big.Int
+	builder := builders[0]
+	switch b := builder.(type) {
+	case *CredentialBuilder:
+		userSecret = b.secret
+	case *DisclosureProofBuilder:
+		userSecret = b.attributes[0]
+	default:
+		return KeyshareResponseRequest[T]{}, nil, errors.New("Unsupported proof builder")
+	}
+
+	challenge, err := builders.ChallengeWithRandomizers(context, nonce, randomizers, signature)
+	if err != nil {
+		return KeyshareResponseRequest[T]{}, nil, err
+	}
+	userResponse := new(big.Int).Add(randomizers["secretkey"], new(big.Int).Mul(challenge, userSecret))
+
+	return KeyshareResponseRequest[T]{
+		Nonce:              nonce,
+		UserResponse:       userResponse,
+		IsSignatureSession: signature,
+		UserChallengeInput: hashInput,
+	}, challenge, nil
+}
+
+func keyshareUserCommitmentsHash[T any](i []KeyshareUserChallengeInput[T]) ([]byte, error) {
+	bts, err := cbor.Marshal(i, cbor.EncOptions{})
+	if err != nil {
+		return nil, err
+	}
+	h := sha256.Sum256(bts)
+	return h[:], nil
+}
+
+// KeyshareResponse generates the keyshare response, using the keyshare secret and the user's
+// input in the keyshare protocol so far.
+func KeyshareResponse[T comparable](
+	secret *big.Int,
+	randomizer *big.Int,
+	req KeyshareCommitmentRequest,
```

Comment:
Something is still a bit off in the naming here. Both `req` and `res` are requests, one for commitment and one for the response phase. So the names `req` and `res` are incoherent